pmd-core/src/main/java/net/sourceforge/pmd/benchmark/TimeTracker.java

```diff
@@ -0,0 +1,255 @@
+        }
+    }
+    
+    /**
+     * An entry in the open timers queue. Defines an operation that has started and hasn't finished yet.
+     */
+    private static class TimerEntry {
+        /* package */ final TimedOperation operation;
+        /* package */ final long start;
+        /* package */ long inNestedOperations = 0;
+        
+        /* package */ TimerEntry(final TimedOperationCategory category, final String label) {
+            this.operation = new TimedOperation(category, label);
+            this.start = System.nanoTime();
+        }
+
+        @Override
+        public String toString() {
+            return "TimerEntry for " + operation;
+        }
+    }
+    
+    /**
+     * Aggregate results measured so far for a given category + label.
+     */
+    /* package */ static class TimedResult {
+        /* package */ AtomicLong totalTime = new AtomicLong();
+        /* package */ AtomicLong selfTime = new AtomicLong();
+        /* package */ AtomicInteger callCount = new AtomicInteger();
+        /* package */ AtomicLong extraDataCounter = new AtomicLong();
+        
+        /**
+         * Adds a new {@link TimerEntry} to the results.
+         * @param timerEntry The entry to be added
+         * @param extraData Any extra data counter to be added
+         * @return The delta time transcurred since the {@link TimerEntry} began.
+         */
+        /* package */ long accumulate(final TimerEntry timerEntry, final long extraData) {
+            final long delta = System.nanoTime() - timerEntry.start;
+            
+            totalTime.getAndAdd(delta);
+            selfTime.getAndAdd(delta - timerEntry.inNestedOperations);
+            callCount.getAndIncrement();
+            extraDataCounter.getAndAdd(extraData);
+            
+            return delta;
+        }
+        
+        /**
+         * Merges the times (and only the times) from another {@link TimedResult} into self.
+         * @param timedResult The {@link TimedResult} to merge
+         */
+        /* package */ void mergeTimes(final TimedResult timedResult) {
+            totalTime.getAndAdd(timedResult.totalTime.get());
+            selfTime.getAndAdd(timedResult.selfTime.get());
+        }
+    }
+    
+    /**
+     * A unique identifier for a timed operation
+     */
+    /* package */ static class TimedOperation {
+        /* package */ final TimedOperationCategory category;
+        /* package */ final String label;
+        
+        /* package */ TimedOperation(final TimedOperationCategory category, final String label) {
+            this.category = category;
+            this.label = label;
+        }
+
+        @Override
+        public int hashCode() {
+            final int prime = 31;
+            int result = 1;
+            result = prime * result + ((category == null) ? 0 : category.hashCode());
+            result = prime * result + ((label == null) ? 0 : label.hashCode());
+            return result;
```

Comment:
Is `Objects.hash(...)` not acceptable? Too slow?

Reply:
this was autogenerated by IDE, just lazy :)